torch/_inductor/comms.py

```diff
@@ -17,61 +20,116 @@
 
 
-def raise_comms(
-    snodes: List[BaseSchedulerNode],
-    node_users: Dict[BaseSchedulerNode, Set[BaseSchedulerNode]],
-    inverse_users: Dict[BaseSchedulerNode, Set[BaseSchedulerNode]],
-) -> List[BaseSchedulerNode]:
+def raise_comms(snodes: List[BaseSchedulerNode], *args) -> List[BaseSchedulerNode]:
     """
-    Greedily moves comms as early as possible (i.e. until we reach an input).
+    Greedily moves comms as early as possible.
     Optimal in terms of communication overlap.
-
-    TODO: We might want to adjust this in the future to account for memory limitations.
-    e.g. when we are compiling FSDP, this heuristics will cause the all-gathers to be prefetched as soon as possible,
-    which is the beginning of the forwards pass. We'll have to either do a special pass for FSDP,
-    or we'll want to redo this pass with memory considerations so we handle the FSDP case in a general way.
     """
-    new_order_reversed: List[BaseSchedulerNode] = []
-    cur_comms: List[BaseSchedulerNode] = []
-    for snode in reversed(snodes):
-        if is_collective(snode.node):
-            cur_comms.append(snode)
-        else:
-            for comm in cur_comms:
-                assert len(inverse_users[comm]) > 0
-            while len(cur_comms) > 0 and any(
-                snode in inverse_users[comm] for comm in cur_comms
-            ):
-                comm = cur_comms.pop(0)
-                new_order_reversed.append(comm)
-            new_order_reversed.append(snode)
-    assert len(cur_comms) <= 1
-    new_order_reversed.extend(tuple_sorted(cur_comms))
-    return new_order_reversed[::-1]
+    return raise_comms_and_sink_waits(snodes, raise_comms=True, sink_waits=False)
+
+
+def raise_comms_and_sink_waits(
+    snodes: List[BaseSchedulerNode],
+    *args,
+    raise_comms: bool = True,
+    sink_waits: bool = True,
+) -> List[BaseSchedulerNode]:
+    # We assign each node a tuple of scores (score_0, score_1, score_2),
+    # decreasing in importance, with a lower value indicating a higher ranking:
+    #
+    # - score_0: the lowest comm_idx among the comm nodes that the node blocks.
+    # If a node doesn't block any comm nodes, its score_0 is set to
+    # sys.maxsize. This score ensures that comm nodes get scheduled as early as
+    # possible.
+    # - score_1: 1 if the node is a wait node, 0 otherwise. This score ensures
+    # that wait nodes are deferred as late as possible.
+    # - score_2: the index of the node in the original topological order. This
+    # score provides stability in case of ties.
+    #
+    # When only raise_comms is True, only score_0 and score_2 are considered.
+    # When only sink_waits is True, only score_1 and score_2 are considered.
+    # When neither is True, the original order is yielded.
+    name_to_snode = {}
+    scores_0, scores_1, scores_2 = {}, {}, {}
+    for idx, snode in enumerate(snodes):
+        for name in snode.get_names():
```

Comment:
To merge to latest master, we might need:
```suggestion
        for name in snode.get_buffer_names():
```